### PR TITLE
Chain Selection Subsystem Logic

### LINKED ISSUE #3103: Stop building on top of blocks that take too long to be approved
Generally a block that takes a long time to be approved will be disputed, but there may be liveness attacks where an adversary interferes with the ability of nodes to recover the available data. If the adversary is performing this type of liveness attack, the block should be ignored by authorship until it is approved. This will be for blocks that are not approved within a few minutes of being imported.

There are two approaches:
  1. Use the blacklisting feature of Substrate. This is most convenient but will cause edge-cases when the block eventually does get approved. We'd have to un-blacklist it in a way that doesn't race with disputes.
  2. Refactor fork-choice rule logic in Substrate to plug in approval-voting to the chain selection system, while still accounting for the BABE counting rules.

### LINKED ISSUE #3235: Fork-choice for parachains based on approvals and disputes
## Problem Statement

One of the goals of parachain consensus is that we never finalize any parachain block which is not considered valid. That means that we have to monitor approval and dispute state and use that to determine not only which blocks we vote to finalize but also which blocks we choose to build on.

We need two rules, one for choosing which block to build on top of, and one for choosing which block to finalize.

**Block Authoring Parent Rule**:
1. Avoid blocks which contain candidates which have lost disputes
2. Avoid blocks which are not approved and are more than S slots old (S = 15 or 90 seconds with 6-second slot times)
3. Choose the best block out of all remaining by comparing BABE score

**Finality Target Rule**:
1. Choose the best block via the **Block Authoring Parent Rule**
2. Avoid blocks which contain candidates which have active disputes
3. Avoid blocks which have not been approved

These two rules, when used in conjunction with each other, will provide the following properties:
1. Nothing disputed will be finalized
2. Chains that are stagnant wil be abandoned
3. Chains that contain invalid candidates will be abandoned
4. Chains which are built as alternatives to abandoned chains will be ordered by BABE rules and can be finalized even when much shorter than the abandoned chain.

**Revert Signals**
We assume that chains which contain invalid candidates eventually issue digests indicating that they should be abandoned as disputes conclude on these chains. These signals take the form `RevertSignal(BlockNumber)`, appearing in the block header, which indicates the ancestor of the signaling block which contains a bad candidate and should be reverted. A block may contain multiple revert signals targeting different ancestors.

----

# Solution

Create a `ForkChoiceSubsystem` which wraps a database and implements the block authoring parent rule. This can be wrapped by a `SelectChain` implementation which will can be plugged into BABE and GRANDPA. The additional constraints on GRANDPA voting can be implemented via the approval voting and dispute coordinator subsystems.

---- 

Data structures:

```rust
struct LeafEntry {
    block_hash: BlockHash,
    babe_score: BabeScore,
}

struct Stagnant(bool);

struct ApprovalStatus {
    // None means approved
    approved: Option<Stagnant>,
    earliest_stagnant_ancestor: Option<BlockHash>,
}

struct RevertStatus {
   // Whether this block has been explicitly reverted by one of its descendants.
   explicitly_reverted: bool,
   // The hash of the earliest unfinalized ancestor which was explicitly reverted.
   earliest_reverted_ancestor: Option<BlockHash>,
}

struct BlockEntry {
    block_hash: BlockHash,
    parent_hash: BlockHash,
    import_time: Timestamp,
    children: Vec<BlockHash>,
    approval_status: ApprovalStatus,
    revert_status: RevertStatus,
    babe_score: BabeScore,
}
```

DB Schema:

```rust
// All unfinalized leaves. If empty, the finalized block is implicitly the fork-choice.
// Sorted descending by BABE score.
Leaves: Vec<LeafEntry>;
// Unfinalized number -> all known block hashes.
BlocksAtHeight: BlockNumber -> Vec<BlockHash>;
// Unfinalized block hash -> entry
BlocksByHash: BlockHash -> Vec<BlockEntry>;
// Import timestamp to block hashes which have not been checked for stagnation.
StagnantQueue: Timestamp -> Vec<BlockHash>;
```

----

### Stagnation

Stagnation refers to blocks which go a long time without being approved. These are typically blocks which contain invalid candidates or the result of bugs in the approval voting subsystems.

A block is stagnant in the fork-choice rule if either
1. Its own approval status is `Some(Stagnant(true))`.
2. It has a stagnant ancestor

We assume that we have two constants:

```rust
// How often to check for stagnation. This should balance frequency with expense.
// When checks are infrequent, validators can have highly varying views of what is stagnant.
// When checks are too frequent, effort is wasted checking with no underlying change.
//
// recommend 5-10 seconds
const STAGNANT_CHECK_INTERVAL: Duration;
// How many slots old a block must be unapproved to be considered stagnant. Recommend 30 slots
const STAGNANT_THRESHOLD: Duration;
```

### Reversion

A block is explicitly reverted if one of its descendants issues a revert signal targeting its block number.
A block is reverted if it is explicitly reverted or is the descendant of an unfinalized explicitly reverted block.

Blocks which contain candidates that have lost disputes are explicitly reverted by the runtime code handling the import of disputes. The descendants of those blocks are implicitly reverted.

### Viable Leaf

A block is viable if:

1. It is not stagnant.
2. It is not reverted.

Note that a block can only be viable if all of its unfinalized ancestors are viable as well.

A block is a viable leaf if:
1. It is viable.
2. It has no descendants which are viable.

Note that it is possible to determine whether a block is viable by inspecting its entry and only its entry.  Furthermore, it is possible to determine parent viability from a child's entry. All iteration is done during updates to the state, while queries to the state are cheap. This is designed to reflect the access patterns: the selection of a block to build upon or finalize should be quick and should not involve any expensive iteration.

-----

### Operations

1. **Import Block**:

Import the entire unknown ancestry of the block, extracting relevant info. Blocks are imported in ascending order. The initial approval state for new blocks should be `Some(not_stagnant)`, with `earliest_stagnant_ancestor` fetched from the parent and `None` if the parent is not in the database. The initial reversion state for new blocks should be `explicitly_reverted = false` and `earliest_reverted_ancestor` based on the parent. 

Add children to any blocks that already exist.  

For each revert signal in any new block, iterate backwards to the referenced ancestor, ignoring if finalized, and set `explicitly_reverted = true`. If necessary, iterate forwards from the ancestor into all descendants and update the `earliest_reverted_ancestor`.

Add each imported block to the `StagnantQueue` map with the slot number at which it will become stagnant - `slot + STAGNANT_THRESHOLD`

Update the set of active leaves:
  * Replace any non-reverted leaves which have been superseded by children
  * Remove any reverted leaves and search recursively starting from the parent of the explicitly reverted blocks.

----

2. **Finalize Block**:

Remove all block entries with number older than the finalized block number.
Remove all block entries with the same number as the finalized block but with different hash, and all of their children.

No entries from `StagnantQueue` are pruned here. They will remain stale but will be cleaned up later by the stagnant detection task.

Remove the finalized block’s entry:

If stagnant, iterate recursively through all descendants and update the `earliest_stagnant_ancestor` as necessary and correct for each block.
If explicitly reverted, iterate recursively through all descendants and update the `earliest_reverted_ancestor` as necessary and correct for each block.

Update the set of active leaves:
* Remove all leaves from orphaned blocks and the finalized block.
* If the finalized block was not viable, then the leaves set must be empty as it can only contain descendants of the finalized block. 
* Therefore: if the active leaves set is not empty, then the finalized block must have been viable. These leaves are still the correct leaves, as no new blocks have been imported, and nothing unfinalized has lost viability.
* Corollary: If the active leaves set is empty, then the finalized block was non-viable and the set should be filled with any viable leaves descending from the finalized block.

It’s also worth noting that the only way a non-viable block can be finalized is if our view in the network vastly differs from the majority due to the slow spread of information, and that typically the finalized block is viable and finalization does not require iteration.

------

3. **Approve Block**:

We assume that the approval voting subsystem sends signals about each approved block, including blocks which are instantaneously approved by virtue of including no candidates or similar protocol constraints.

The approved block should be loaded and its approval status should change to `None`. If the approved block was stagnant and the `earliest_stagnant_ancestor` of the approved block is `None`, then descendants should be visited recursively and have their `earliest_stagnant_ancestor` updated accordingly.

Update the set of active leaves:
* If the approved block was not stagnant, then no block has changed in viability and no change in the viable leaves should be made.
* If the approved block's parent was not viable then no block has changed in viability and no change in the viable leaves should be made.
* If the approved block's parent was viable and the approved block has now become viable, then its parent should be removed from the viable leaves set, and the approved block and its descendants should be searched for viable leaves. This can be combined with the `earliest_stagnant_ancestor` update of descendants.

------

4. **Detect Stagnant**:

This is invoked every `STAGNANT_CHECK_INTERVAL` time. Iterate and drain in ascending order every entry in `StagnantQueue` up to the current timestamp.

For each block hash there are 3 possibilities:
1. There is no entry for the block. Continue; this is an orphaned block.
2. There is an entry for the block and it is approved. Continue; there is no change to be made
3. There is an entry for the block and it is not approved. Mark as stagnant. If `earliest_stagnant_ancestor` is `None`, iterate all descendants and set `earliest_stagnant_ancestor`.

Update the set of active leaves:
* If the block was viable and is now non-viable, then any leaves that were encountered while updating `earliest_stagnant_ancestor` should be removed, and the parent of the block should, if unfinalized, be added to the active leaves set.
* If the block has not changed in viability, then no leaf has changed in viability.
* Generally, since we are starting with earlier blocks as opposed to later ones, that only the first set of stagnant blocks should involve any significant change to the active leaves set.

### PR DESCRIPTION
**Pull request by rphmeier**
Closes #3235 
Closes #3103 

This implements the logic of the chain selection subsystem as described in #3262 . I took a slightly different approach to DB handling than what is done in other subsystems that wrap a DB, by using a `Backend` trait and defining `BackendWriteOps`, along with a helper struct `BackendOverlay` for maintaining a consistent state when querying changes that haven't been committed. I think this simplifies the code and will make it more difficult to write logic errors in the class of #3311 

The primary code here is located within the `tree` submodule which is used to maintain a coherent view of all unfinalized blocks and implements the core operations of import, finalization, reversion, approval, and stagnation.

Unlike #3235 which describes separate fields for stagnation and reversion, I realized that I could combine these fields into a single field for tracking the earliest unviable ancestor of each block. This allows for further code reduction as we don't need to handle these in separate paths. The helper function for `propagate_viability_update` does most of the heavy lifting and the algorithm functions the same both when propagating a viable->unviable change and unviable->viable changes.

TODO:
  - [x] Block import
  - [x] Block reversion logs
  - [x] Block finalization
  - [x] Approve Message
  - [x] Stagnant updates
  - [x] Tests

Deferred to later:
  - Plug into overseer
  - Implement DB
  - Send Approve messages from Approval Voting (even when insta-approving)
  - Stagnation loop

### CODE REVIEW

File: node/core/chain-selection/src/lib.rs
Comment (Lldenaurois) [inline node/core/chain-selection/src/lib.rs:50]:
```suggestion
		matches!(self, &Approval::Stagnant)
```

File: node/core/chain-selection/src/lib.rs
Comment (Lldenaurois) [inline node/core/chain-selection/src/lib.rs:79]: The Implementer's Guide mentions the first criteria to be:

`It [the viable block] is or descends from the finalized block`

Do we need to add an additional check to ensure that some parent is the finalized block?

**Review by Lldenaurois — COMMENTED** [review]
First pass

File: node/core/chain-selection/src/lib.rs
Comment (drahnr) [inline node/core/chain-selection/src/lib.rs:63]: Comment is not entirely clear, what would `approval` be set to in that case?

File: node/core/chain-selection/src/lib.rs
Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:79]: I see what you mean. One goal of this implementation is to ensure that only blocks descending from the finalized chain are kept in the DB.

File: node/core/chain-selection/src/lib.rs
Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:63]: outdated comment when `Approval` was an `Option<bool>`

**Review by drahnr — COMMENTED** [review]
First pass LGTM, I need to give the logic another pass to be certain

File: node/core/chain-selection/src/lib.rs
Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:169]: For our use-case of detecting stagnation in 90 seconds we shouldn't need to worry about whether the timestamp is perfectly accurate or even monotonic.

Comment (drahnr) [reply node/core/chain-selection/src/lib.rs:176]: My concern was mostly when non-monotonic comes into play, and the time advances unexpectedly. Then the timeout detection triggers earlier (technically could even instantly). If this is not something to consider (i.e. summer to wintertime change with system clock change) then leave as is

Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:176]: I see what you mean, that the timeout could trigger instantly. I'm not very familiar with the internal workings of the clock but I imagine if a clock is slipping by 90 seconds something is seriously wrong. I would expect slippage of no more than a few seconds and typically less, and due to network asynchrony validators are expected to have slightly different views of when blocks become stagnant anyway.

Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:176]: I can add a comment to address the concern about non-monotonicity, if that would help.

Comment (drahnr) [reply node/core/chain-selection/src/lib.rs:184]: Usually this is a more windows centric issue, where the OS happily messes with the system time.

A simple one liner comment would fine, since we don't currently target windows as a supported OS.

**Review by ordian — COMMENTED** [review]
some nits from the first pass

File: node/core/chain-selection/src/tree.rs
Comment (rphmeier) [reply node/core/chain-selection/src/tree.rs]: (removed)

File: node/network/bridge/src/tests.rs
Comment (rphmeier) [reply node/network/bridge/src/tests.rs:1260]: 😱

File: node/core/chain-selection/src/backend.rs
Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:96]: I assume the underlying backend has some caching, which rocksdb does, so this is OK.

Try implementing `fn propagate_viability_update` without the overlay if you don't think it's necessary!

Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:96]: If we'd like to cache on load, and find that rocksdb cache is not sufficient, I think that the best approach to take would be to use trait composability.

So we'd write a 

```rust
struct CachedBackend<B> { cache, inner: B }

impl<B: Backend> Backend for CachedBackend<B> { ... }
```

This would allow us to implement a targeted cache without affecting the logic of the overlay.

Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:96]: I actually initially tried to implement `Backend for OverlayedBackend<B: Backend>` but the `first_block_number` and `stagnant_at_up_to` methods proved pretty much impossible.

**Comment by rphmeier** [comment]
I've tested this to a satisfactory degree. I think the tests are very comprehensive, at least for the features we support now. The follow-up PR will add stagnation and with that we'll be able to test that approvals make chains viable again. I found a few bugs but did not find any bugs in `tree::propagate_viability_update`.

File: node/core/chain-selection/src/lib.rs
Comment (ordian) [inline node/core/chain-selection/src/lib.rs:106]: I'd suggest to use `!maches!(ord, std::cmp::Ordering::Equal)` if we want to merge it before our CI images are updated to rust 1.53.

File: node/core/chain-selection/src/backend.rs
Comment (ordian) [reply node/core/chain-selection/src/backend.rs:96]: Ok, fair enough. I don't think it's worth blocking this PR on this concern, but it's worth revisiting once we merge all the dbs.

File: node/core/chain-selection/src/backend.rs
Comment (ordian) [inline node/core/chain-selection/src/backend.rs:206]: should we put a cap in this function, as is it guaranteed that it will not go to genesis?

File: node/core/chain-selection/src/backend.rs
Comment (ordian) [inline node/core/chain-selection/src/backend.rs:224]: a note, that this function is `O(distance(leaf, required)^2)`, might be heavy if too many leaves

**Review by ordian — APPROVED** [review]
I didn't understand all of it, but further improvements can go to https://github.com/paritytech/polkadot/issues/3293.

File: node/core/chain-selection/src/backend.rs
Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:206]: No, that's not necessary IMO. It will never iterate beyond the finalized block

File: node/core/chain-selection/src/backend.rs
Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:224]: The existing implementation of `best_leaf_containing` that we use from the client is the same complexity, but I will add a comment

**Review by drahnr — APPROVED** [review]
A few small nits and notes, as far as I can see follows `protocol-chain-selection.md` exactly :+1:

File: node/core/chain-selection/src/backend.rs
Comment (drahnr) [inline node/core/chain-selection/src/backend.rs:191]: This could be part of the `Backend`.

File: node/core/chain-selection/src/lib.rs
Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:200]: I see. This is a pattern we use elsewhere as well, so that should be a separate batch refactoring

File: node/core/chain-selection/src/lib.rs
Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:231]: (done in #3328 )

File: node/core/chain-selection/src/lib.rs
Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:481]: Yeah, this is kind of similar to the problem of #3311 - I tried for a few hours to create a `locksmith` crate with an API we could use to ensure write-ops are committed. But this is minor and right now our usage is correct.

File: node/core/chain-selection/src/backend.rs
Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:191]: Yes, I can see that, but I believe it's simpler right now to expose a minimal `Backend` API and build on top of that. We're going to be disk-bound by this anyway.

File: node/core/chain-selection/src/lib.rs
Comment (rphmeier) [reply node/core/chain-selection/src/lib.rs:118]: Yeah, at first I was doing some iterator magic to find the position and it became unreadable quickly

**Comment by rphmeier** [comment]
bot merge

File: node/core/chain-selection/src/backend.rs
Comment (Lldenaurois) [reply node/core/chain-selection/src/backend.rs:191]: This function is O(n), so in the worst case someone would pass the genesis hash in and we would basically read the entire chain, right?

I wonder whether it makes sense to put a limit on the depth of this loop...

File: node/core/chain-selection/src/lib.rs
Comment (Lldenaurois) [reply node/core/chain-selection/src/lib.rs:200]: Filing an issue.

File: node/core/chain-selection/src/backend.rs
Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:191]: @Lldenaurois It's actually O(block_height - finalized_height) because we only store unfinalized subtrees here.

Comment (rphmeier) [reply node/core/chain-selection/src/backend.rs:191]: Right now it's more inefficient than it needs to be but we could easily keep a `HashSet<Hash>` of block-entries we've visited already and bail early when we encounter a parent hash we've already seen. Most chains will have some common ancestor before the last finalized block.

**Pull request merged**
